docs/mmds/mmds-user-guide.md

```diff
@@ -0,0 +1,240 @@
+
+### Example
+
+```bash
+MMDS_IPV4_ADDR=169.254.170.2
+MMDS_NET_IF=eth0
+ip route add ${MMDS_IPV4_ADDR} dev ${MMDS_NET_IF}
+```
+
+# Inserting and updating metadata
+
+Inserting and updating metadata is possible through the firecracker API server.
+These operations can be issued before and after the microVM is running. To
+insert metadata into MMDS, a HTTP `PUT` request to the `/mmds` resource has to
+be issued. This request must have a payload with metadata structured in 
+[JSON](https://tools.ietf.org/html/rfc7159) format. To replace existing
+metadata, a subsequent HTTP `PUT` request to the `/mmds` resource must be
+issued, using as a payload the new metadata. A complete description of metadata
+insertion firecracker API can be found in the
+[firecracker swagger file](../../src/api_server/swagger/firecracker.yaml).
+
+
+### Example
+
+```bash
+curl --unix-socket /tmp/firecracker.socket -i \
+    -X PUT "http://localhost/mmds" \
+    -H "accept: application/json" \
+    -H "Content-Type: application/json" \
+    -d "{
+            "latest": {
+                  "meta-data": {
+                       "ami-id": "ami-12345678",
+                       "reservation-id": "r-fea54097",
+                       "local-hostname": "ip-10-251-50-12.ec2.internal",
+                       "public-hostname": "ec2-203-0-113-25.compute-1.amazonaws.com",
+                       "network": {
+                            "interfaces": {
+                                 "macs": {
+                                      "02:29:96:8f:6a:2d": {
+                                           "device-number": "13345342",
+                                           "local-hostname": "localhost",
+                                           "subnet-id": "subnet-be9b61d"
+                                      }
+                                 }
+                            }
+                       }
+                  }
+            }      
+    }"
+```
+
+To partially update existing metadata, a HTTP `PATCH` request to the `/mmds`
+resource has to be issued, using as a payload the metadata patch, as
+[JSON Merge Patch]((https://tools.ietf.org/html/rfc7396)) functionality
+describes. A complete description of updating metadata firecracker API can be
+found in the [firecracker swagger file](../../src/api_server/swagger/firecracker.yaml).
+
+### Example
+
+```bash
+curl --unix-socket /tmp/firecracker.socket -i \
+    -X PATCH "http://localhost/mmds" \
+    -H "accept: application/json" \
+    -H "Content-Type: application/json" \
+    -d "{
+            "latest": {
+                  "meta-data": {
+                       "ami-id": "ami-87654321",
+                       "reservation-id": "r-79054aef",
+                  }
+            }      
+    }"
+```
+
+# Retrieving metadata
+
+MicroVM metadata can be retrieved:
+* from host operating system, communicating with the firecracker API server.
+* from guest operating system, communicating directly with MMDS.
```

Comment:
communitating directly with MMDS -> not clear can you also do PUT from microVM, or we accept just GET.
I would rephrase this to something like - from guest OS doing HTTP GET operations to MMDS IP with query ....
Also from host operating system, doing HTTP queries to firecracker API server at  /mmds endpoint.

Reply:
This list of ways to retrieve metadata was intended to be a summary of the routes followed by queries, but not actually describing how these queries/requests look like. Related to how this communication looks like, there are two chapters which follow, for HOST and GUEST, with API/requests examples for retrieving metadata from MMDS. If I would rephrase how you suggest, this would mean that I would repeat the same information in the following chapters, because they intend to describe the API and provide examples.

Reply:
I would be specific on what HTTP methods can be used and from where